{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(291, 483817) (291,)\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "x_file_path = './CellTOSG/brain_sc_output/processed_data/brain/alzheimer\\'s_disease/alzheimer\\'s_disease_X_partition_0.npy'\n",
    "y_file_path = './CellTOSG/brain_sc_output/processed_data/brain/alzheimer\\'s_disease/alzheimer\\'s_disease_Y_partition_0.npy'\n",
    "\n",
    "x = np.load(x_file_path)\n",
    "y = np.load(y_file_path)\n",
    "print(x.shape, y.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2, 33349084) (483817, 1)\n",
      "[[     0      1      2 ... 483091 483091 483091]\n",
      " [278326 278327 278328 ... 386373 398848 437378]]\n"
     ]
    }
   ],
   "source": [
    "edge_index_path = './CellTOSG/brain_sc_output/edge_index.npy'\n",
    "x_desc_path = './CellTOSG/brain_sc_output/X_descriptions.npy'\n",
    "\n",
    "edge_index = np.load(edge_index_path)\n",
    "x_desc = np.load(x_desc_path, allow_pickle=True)\n",
    "print(edge_index.shape, x_desc.shape)\n",
    "print(edge_index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['GTP-binding protein involved in protein trafficking; may modulate vesicle budding and uncoating within the Golgi apparatus. {ECO:0000250|UniProtKB:P84084}.;  (Microbial infection) Functions as an allosteric activator of the cholera toxin catalytic subunit, an ADP-ribosyltransferase. {ECO:0000305}.'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_desc[278326]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert x_desc to list of strings\n",
    "x_desc = [str(i) for i in x_desc]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[\"['ADP ribosylation factor 5 [Source:HGNC Symbol;Acc:HGNC:658]']\",\n",
       " \"['mannose-6-phosphate receptor, cation dependent [Source:HGNC Symbol;Acc:HGNC:6752]']\"]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_desc[:2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Name emb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(483817, 4)\n",
      "(278326, 4) (205491, 4)\n"
     ]
    }
   ],
   "source": [
    "transcript_name_emb_path = './BioMedGraphica/embeddings/Entity/transcript_embeddings.npy'\n",
    "protein_name_emb_path = './BioMedGraphica/embeddings/Entity/protein_embeddings.npy'\n",
    "\n",
    "transcript_name_emb = np.load(transcript_name_emb_path)\n",
    "protein_name_emb = np.load(protein_name_emb_path)\n",
    "name_emb = np.concatenate((transcript_name_emb, protein_name_emb), axis=0)\n",
    "print(name_emb.shape)\n",
    "print(transcript_name_emb.shape, protein_name_emb.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import torch\n",
    "import matplotlib.pyplot as plt\n",
    "import torch.nn.functional as F\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from torch_geometric.nn import GATConv\n",
    "from torch_geometric.data import Data\n",
    "from transformers import AutoTokenizer, AutoModel\n",
    "from tqdm import tqdm\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from typing import List, Tuple, Dict\n",
    "from pykeen.pipeline import pipeline\n",
    "from pykeen.triples import TriplesFactory\n",
    "from torch.utils.data import DataLoader, Dataset\n",
    "from pykeen.training.callbacks import TrainingCallback  # Updated base class for callbacks\n",
    "\n",
    "from pykeen.models import TransR\n",
    "from pykeen.training import SLCWATrainingLoop\n",
    "\n",
    "from torch_geometric.nn import TransformerConv\n",
    "\n",
    "def prepare_sentences(data: pd.DataFrame) -> Tuple[List[str], List[str]]:\n",
    "    \"\"\"\n",
    "    Convert each row of the data into a formatted sentence and return IDs.\n",
    "\n",
    "    Args:\n",
    "        data (pd.DataFrame): DataFrame containing the data.\n",
    "\n",
    "    Returns:\n",
    "        Tuple[List[str], List[str]]: List of sentences and list of IDs.\n",
    "    \"\"\"\n",
    "    # Generate the formatted sentences\n",
    "    sentences = [\n",
    "        \"; \".join(str(value) for key, value in row.items() if key != \"BioMedGraphica_ID\" and pd.notna(value))\n",
    "        for row in data.to_dict(orient=\"records\")\n",
    "    ]\n",
    "    ids = data[data.columns[0]].tolist()\n",
    "    return sentences, ids\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\hemingzhang\\AppData\\Local\\Temp\\ipykernel_25784\\1567783997.py:2: DtypeWarning: Columns (1,2,5,7) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  protein_language_name_id_df = pd.read_csv('./BioMedGraphica/Entity\\Protein/biomedgraphica_protein_language_name_id.csv')\n"
     ]
    }
   ],
   "source": [
    "transcript_language_name_id_df = pd.read_csv('./BioMedGraphica/Entity/Transcript/biomedgraphica_transcript_language_name_id.csv')\n",
    "protein_language_name_id_df = pd.read_csv('./BioMedGraphica/Entity\\Protein/biomedgraphica_protein_language_name_id.csv')\n",
    "\n",
    "transcript_name_sentence, transcript_ids = prepare_sentences(transcript_language_name_id_df)\n",
    "protein_name_sentence, protein_ids = prepare_sentences(protein_language_name_id_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioMedGraphica_ID</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BMG_TS000001</td>\n",
       "      <td>ENST00000000233; ENST00000000233.10; ENSG00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BMG_TS000002</td>\n",
       "      <td>ENST00000000412; ENST00000000412.8; ENSG000000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BMG_TS000003</td>\n",
       "      <td>ENST00000000442; ENST00000000442.11; ENSG00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BMG_TS000004</td>\n",
       "      <td>ENST00000001008; ENST00000001008.6; ENSG000000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BMG_TS000005</td>\n",
       "      <td>ENST00000001146; ENST00000001146.7; ENSG000000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483812</th>\n",
       "      <td>BMG_PT205487</td>\n",
       "      <td>UniProt ID:X6RGR3; RefSeq ID:NP_001353609.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483813</th>\n",
       "      <td>BMG_PT205488</td>\n",
       "      <td>UniProt ID:X6RLR1; RefSeq ID:NP_001268356.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483814</th>\n",
       "      <td>BMG_PT205489</td>\n",
       "      <td>UniProt ID:X6RLX0; RefSeq ID:XP_054227478.1; R...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483815</th>\n",
       "      <td>BMG_PT205490</td>\n",
       "      <td>UniProt ID:X6RM00; RefSeq ID:XP_047284540.1; R...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483816</th>\n",
       "      <td>BMG_PT205491</td>\n",
       "      <td>UniProt ID:X6RM59; RefSeq ID:NP_001002010.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>483817 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       BioMedGraphica_ID                                               Name\n",
       "0           BMG_TS000001  ENST00000000233; ENST00000000233.10; ENSG00000...\n",
       "1           BMG_TS000002  ENST00000000412; ENST00000000412.8; ENSG000000...\n",
       "2           BMG_TS000003  ENST00000000442; ENST00000000442.11; ENSG00000...\n",
       "3           BMG_TS000004  ENST00000001008; ENST00000001008.6; ENSG000000...\n",
       "4           BMG_TS000005  ENST00000001146; ENST00000001146.7; ENSG000000...\n",
       "...                  ...                                                ...\n",
       "483812      BMG_PT205487        UniProt ID:X6RGR3; RefSeq ID:NP_001353609.1\n",
       "483813      BMG_PT205488        UniProt ID:X6RLR1; RefSeq ID:NP_001268356.1\n",
       "483814      BMG_PT205489  UniProt ID:X6RLX0; RefSeq ID:XP_054227478.1; R...\n",
       "483815      BMG_PT205490  UniProt ID:X6RM00; RefSeq ID:XP_047284540.1; R...\n",
       "483816      BMG_PT205491        UniProt ID:X6RM59; RefSeq ID:NP_001002010.2\n",
       "\n",
       "[483817 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "name_sentence = transcript_name_sentence + protein_name_sentence\n",
    "name_ids = transcript_ids + protein_ids\n",
    "\n",
    "# construct the s_name.csv\n",
    "name_df = pd.DataFrame({'BioMedGraphica_ID': name_ids, 'Name': name_sentence})\n",
    "name_df.to_csv('./CellTOSG/s_name.csv', index=False)\n",
    "display(name_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Desc emb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import List, Tuple, Dict\n",
    "from torch.utils.data import DataLoader, Dataset\n",
    "\n",
    "class SentenceDataset(Dataset):\n",
    "    def __init__(self, sentences: List[str]):\n",
    "        self.sentences = sentences\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.sentences)\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        return self.sentences[idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\hemingzhang\\AppData\\Local\\anaconda3\\envs\\mkg\\lib\\site-packages\\tqdm\\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import torch\n",
    "import matplotlib.pyplot as plt\n",
    "import torch.nn.functional as F\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from torch_geometric.nn import GATConv\n",
    "from torch_geometric.data import Data\n",
    "from transformers import AutoTokenizer, AutoModel\n",
    "from tqdm import tqdm\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from typing import List, Tuple, Dict\n",
    "from pykeen.pipeline import pipeline\n",
    "from pykeen.triples import TriplesFactory\n",
    "from torch.utils.data import DataLoader, Dataset\n",
    "from pykeen.training.callbacks import TrainingCallback  # Updated base class for callbacks\n",
    "\n",
    "from pykeen.models import TransR\n",
    "from pykeen.training import SLCWATrainingLoop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class TextEmb():\n",
    "    def __init__(self, model_path: str = \"microsoft/deberta-v3-small\", device: str = \"cuda\"):\n",
    "        \"\"\"\n",
    "        Args:\n",
    "            model_path (str, optional): Path to the deberta model. Defaults to 'microsoft/deberta-v3-small'.\n",
    "            device (str, optional): Device to run the model on ('cpu' or 'cuda'). Defaults to 'cpu'.\n",
    "        \"\"\"\n",
    "        self.model_path = model_path\n",
    "        self.device = device\n",
    "        self.model = None\n",
    "        self.tokenizer = None\n",
    "\n",
    "    def load_model(self):\n",
    "        \"\"\"\n",
    "        Load the deberta model and tokenizer from the specified model path.\n",
    "        \"\"\"\n",
    "        self.tokenizer = AutoTokenizer.from_pretrained(self.model_path)\n",
    "        self.model = AutoModel.from_pretrained(self.model_path).to(self.device)\n",
    "\n",
    "    def generate_embeddings(self, sentences: List[str], batch_size: int = 32, text_emb_dim: int = 64) -> List[float]:\n",
    "        \"\"\"\n",
    "        Generate a single-dimensional embedding for each sentence.\n",
    "\n",
    "        Args:\n",
    "            sentences (List[str]): List of sentences to embed.\n",
    "            batch_size (int, optional): Batch size for DataLoader. Defaults to 32.\n",
    "\n",
    "        Returns:\n",
    "            List[float]: List of single-dimensional embeddings.\n",
    "        \"\"\"\n",
    "        dataset = SentenceDataset(sentences)\n",
    "        dataloader = DataLoader(dataset, batch_size=batch_size, shuffle=False)\n",
    "\n",
    "        embeddings = []\n",
    "        for batch in tqdm(dataloader, desc=\"Embedding sentences\", unit=\"batch\"):\n",
    "            inputs = self.tokenizer(batch, return_tensors=\"pt\", padding=True, truncation=True, max_length=512).to(self.device)\n",
    "            with torch.no_grad():\n",
    "                outputs = self.model(**inputs)\n",
    "            batch_embeddings = torch.mean(outputs.last_hidden_state, dim=1).squeeze().cpu()\n",
    "            # Adaptive pooling to 64 dimensions (pooling over the hidden dimension)\n",
    "            projected = torch.nn.functional.adaptive_avg_pool1d(batch_embeddings.unsqueeze(1), output_size=text_emb_dim).squeeze(1)\n",
    "            embeddings.extend(projected.tolist())\n",
    "        return embeddings\n",
    "\n",
    "    def save_embeddings(self, embeddings: List[float], output_npy_path: str) -> None:\n",
    "        \"\"\"\n",
    "        Save embeddings to a .npy file and IDs to a CSV file.\n",
    "\n",
    "        Args:\n",
    "            embeddings (List[float]): List of single-dimensional embeddings.\n",
    "            ids (List[str]): List of corresponding IDs.\n",
    "            output_npy_path (str): Path to save the .npy file.\n",
    "            output_csv_path (str): Path to save the index CSV file.\n",
    "        \"\"\"\n",
    "        # Save the embeddings to .npy file\n",
    "        np.save(output_npy_path, np.array(embeddings))\n",
    "        print(f\"Embeddings saved at {output_npy_path} with shape {np.array(embeddings).shape}\")\n",
    "\n",
    "    def process_data_and_generate_embeddings(self, sentences: List, batch_size: str, text_emb_dim: int) -> Tuple[List[float], List[str]]:\n",
    "\n",
    "        embeddings = self.generate_embeddings(sentences, batch_size, text_emb_dim)\n",
    "        # print generated embeddings at shape\n",
    "        print(f\"Generated embeddings at shape {np.array(embeddings).shape}\")\n",
    "        return embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\imhem\\AppData\\Local\\Temp\\ipykernel_30944\\473054218.py:1: DtypeWarning: Columns (1,2) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  mapping_table = pd.read_csv('./CellTOG/brain_sc_output/mapping_table.csv')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>original_id</th>\n",
       "      <th>original_index</th>\n",
       "      <th>BioMedGraphica_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>BMG_TS000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>M6PR</td>\n",
       "      <td>M6PR</td>\n",
       "      <td>BMG_TS000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ESRRA</td>\n",
       "      <td>ESRRA</td>\n",
       "      <td>BMG_TS000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>BMG_TS000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BMG_TS000005</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index original_id original_index BioMedGraphica_ID\n",
       "0      0        ARF5           ARF5      BMG_TS000001\n",
       "1      1        M6PR           M6PR      BMG_TS000002\n",
       "2      2       ESRRA          ESRRA      BMG_TS000003\n",
       "3      3       FKBP4          FKBP4      BMG_TS000004\n",
       "4      4         NaN            NaN      BMG_TS000005"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "483817\n",
      "['BMG_TS000001', 'BMG_TS000002']\n"
     ]
    }
   ],
   "source": [
    "mapping_table = pd.read_csv('./CellTOSG/brain_sc_output/mapping_table.csv')\n",
    "display(mapping_table.head())\n",
    "ids = mapping_table['BioMedGraphica_ID'].tolist()\n",
    "print(len(ids))\n",
    "print(ids[:2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_path = \"microsoft/deberta-v3-small\"\n",
    "device = \"cuda\"\n",
    "batch_size = 128\n",
    "desc_emb_dim = 1\n",
    "output_npy_path = \"./CellTOSG/x_desc_emb.npy\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\imhem\\anaconda3\\envs\\mkg\\Lib\\site-packages\\transformers\\convert_slow_tokenizer.py:561: UserWarning: The sentencepiece tokenizer that you are converting to a fast tokenizer uses the byte fallback option which is not implemented in the fast tokenizers. In practice this means that the fast version of the tokenizer can produce unknown tokens whereas the sentencepiece version would have converted these unknown tokens into a sequence of byte tokens matching the original piece of text.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "encoder = TextEmb(model_path, device)\n",
    "encoder.load_model()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "embeddings = encoder.process_data_and_generate_embeddings(x_desc, batch_size, desc_emb_dim)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "encoder.save_embeddings(embeddings, output_npy_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioMedGraphica_ID</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BMG_TS000001</td>\n",
       "      <td>ADP ribosylation factor 5 [Source:HGNC Symbol;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BMG_TS000002</td>\n",
       "      <td>mannose-6-phosphate receptor, cation dependent...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BMG_TS000003</td>\n",
       "      <td>estrogen related receptor alpha [Source:HGNC S...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BMG_TS000004</td>\n",
       "      <td>FKBP prolyl isomerase 4 [Source:HGNC Symbol;Ac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BMG_TS000005</td>\n",
       "      <td>cytochrome P450 family 26 subfamily B member 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205486</th>\n",
       "      <td>BMG_PT205487</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205487</th>\n",
       "      <td>BMG_PT205488</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205488</th>\n",
       "      <td>BMG_PT205489</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205489</th>\n",
       "      <td>BMG_PT205490</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205490</th>\n",
       "      <td>BMG_PT205491</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>482461 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       BioMedGraphica_ID                                        Description\n",
       "0           BMG_TS000001  ADP ribosylation factor 5 [Source:HGNC Symbol;...\n",
       "1           BMG_TS000002  mannose-6-phosphate receptor, cation dependent...\n",
       "2           BMG_TS000003  estrogen related receptor alpha [Source:HGNC S...\n",
       "3           BMG_TS000004  FKBP prolyl isomerase 4 [Source:HGNC Symbol;Ac...\n",
       "4           BMG_TS000005  cytochrome P450 family 26 subfamily B member 1...\n",
       "...                  ...                                                ...\n",
       "205486      BMG_PT205487                                                NaN\n",
       "205487      BMG_PT205488                                                NaN\n",
       "205488      BMG_PT205489                                                NaN\n",
       "205489      BMG_PT205490                                                NaN\n",
       "205490      BMG_PT205491                                                NaN\n",
       "\n",
       "[482461 rows x 2 columns]"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_desc_df = pd.read_csv('./BioMedGraphica/Entity/Transcript/biomedgraphica_transcript_description.csv')\n",
    "protein_desc_df = pd.read_csv('./BioMedGraphica/Entity/Protein/biomedgraphica_protein_description.csv').rename(columns={'UniProt': 'Description'})\n",
    "desc_df = pd.concat([transcript_desc_df, protein_desc_df], axis=0)\n",
    "desc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\hemingzhang\\AppData\\Local\\Temp\\ipykernel_25784\\2600793688.py:2: DtypeWarning: Columns (1,2,5,7) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  protein_df = pd.read_csv('./BioMedGraphica/Entity/Protein/biomedgraphica_protein.csv')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioMedGraphica_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BMG_TS000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BMG_TS000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BMG_TS000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BMG_TS000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BMG_TS000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205486</th>\n",
       "      <td>BMG_PT205487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205487</th>\n",
       "      <td>BMG_PT205488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205488</th>\n",
       "      <td>BMG_PT205489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205489</th>\n",
       "      <td>BMG_PT205490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205490</th>\n",
       "      <td>BMG_PT205491</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>483817 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       BioMedGraphica_ID\n",
       "0           BMG_TS000001\n",
       "1           BMG_TS000002\n",
       "2           BMG_TS000003\n",
       "3           BMG_TS000004\n",
       "4           BMG_TS000005\n",
       "...                  ...\n",
       "205486      BMG_PT205487\n",
       "205487      BMG_PT205488\n",
       "205488      BMG_PT205489\n",
       "205489      BMG_PT205490\n",
       "205490      BMG_PT205491\n",
       "\n",
       "[483817 rows x 1 columns]"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_df = pd.read_csv('./BioMedGraphica/Entity/Transcript/biomedgraphica_transcript.csv')\n",
    "protein_df = pd.read_csv('./BioMedGraphica/Entity/Protein/biomedgraphica_protein.csv')\n",
    "x_df = pd.concat([transcript_df, protein_df], axis=0)[['BioMedGraphica_ID']]\n",
    "x_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioMedGraphica_ID</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BMG_TS000001</td>\n",
       "      <td>ADP ribosylation factor 5 [Source:HGNC Symbol;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BMG_TS000002</td>\n",
       "      <td>mannose-6-phosphate receptor, cation dependent...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BMG_TS000003</td>\n",
       "      <td>estrogen related receptor alpha [Source:HGNC S...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BMG_TS000004</td>\n",
       "      <td>FKBP prolyl isomerase 4 [Source:HGNC Symbol;Ac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BMG_TS000005</td>\n",
       "      <td>cytochrome P450 family 26 subfamily B member 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483812</th>\n",
       "      <td>BMG_PT205487</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483813</th>\n",
       "      <td>BMG_PT205488</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483814</th>\n",
       "      <td>BMG_PT205489</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483815</th>\n",
       "      <td>BMG_PT205490</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483816</th>\n",
       "      <td>BMG_PT205491</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>483817 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       BioMedGraphica_ID                                        Description\n",
       "0           BMG_TS000001  ADP ribosylation factor 5 [Source:HGNC Symbol;...\n",
       "1           BMG_TS000002  mannose-6-phosphate receptor, cation dependent...\n",
       "2           BMG_TS000003  estrogen related receptor alpha [Source:HGNC S...\n",
       "3           BMG_TS000004  FKBP prolyl isomerase 4 [Source:HGNC Symbol;Ac...\n",
       "4           BMG_TS000005  cytochrome P450 family 26 subfamily B member 1...\n",
       "...                  ...                                                ...\n",
       "483812      BMG_PT205487                                                NaN\n",
       "483813      BMG_PT205488                                                NaN\n",
       "483814      BMG_PT205489                                                NaN\n",
       "483815      BMG_PT205490                                                NaN\n",
       "483816      BMG_PT205491                                                NaN\n",
       "\n",
       "[483817 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "s_desc_df = pd.merge(x_df, desc_df, on='BioMedGraphica_ID', how='left')\n",
    "s_desc_df = s_desc_df.reset_index(drop=True)[['BioMedGraphica_ID', 'Description']]\n",
    "display(s_desc_df)\n",
    "s_desc_df.to_csv('./CellTOSG/s_desc.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Seq emb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_seq_path = './BioMedGraphica/Entity/Transcript/biomedgraphica_transcript_biochem.csv'\n",
    "protein_seq_path = './BioMedGraphica/Entity/Protein/biomedgraphica_protein_biochem.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioMedGraphica_ID</th>\n",
       "      <th>transcript_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BMG_TS000001</td>\n",
       "      <td>CUGCUGCUGCUGCGCCCCAUCCCCCCGCGGCCGGCCAGUUCCAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BMG_TS000002</td>\n",
       "      <td>AGAGUGGGGCACAGCGAGGCGCUAGGGGGAACGCUGGCCUCUGAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BMG_TS000003</td>\n",
       "      <td>GUCAGCUGGAGGAAGCGGAGUAGGAAGCGGCCGCGAUGUCCUUUUG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BMG_TS000004</td>\n",
       "      <td>CCUACCCCAGCUCUCGCGCCGCGUGCAGAGGUGCUCAAGCCUCCUC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BMG_TS000005</td>\n",
       "      <td>ACAGCCAAUCCCCCGAGCGGCCGCCAACAUGCUCUUUGAGGGCUUG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  BioMedGraphica_ID                                transcript_sequence\n",
       "0      BMG_TS000001  CUGCUGCUGCUGCGCCCCAUCCCCCCGCGGCCGGCCAGUUCCAGCC...\n",
       "1      BMG_TS000002  AGAGUGGGGCACAGCGAGGCGCUAGGGGGAACGCUGGCCUCUGAAA...\n",
       "2      BMG_TS000003  GUCAGCUGGAGGAAGCGGAGUAGGAAGCGGCCGCGAUGUCCUUUUG...\n",
       "3      BMG_TS000004  CCUACCCCAGCUCUCGCGCCGCGUGCAGAGGUGCUCAAGCCUCCUC...\n",
       "4      BMG_TS000005  ACAGCCAAUCCCCCGAGCGGCCGCCAACAUGCUCUUUGAGGGCUUG..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "transcript_seq_df = pd.read_csv(transcript_seq_path)\n",
    "display(transcript_seq_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(278326, 2)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_seq_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_sequences = transcript_seq_df['transcript_sequence'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import torch\n",
    "from tqdm import tqdm\n",
    "from typing import List\n",
    "from dna_gpt.dna_gpt import DNAGPT\n",
    "from dna_gpt.tokenizer import KmerTokenizer\n",
    "from transformers import GPT2LMHeadModel, GPT2Tokenizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import torch\n",
    "import matplotlib.pyplot as plt\n",
    "import torch.nn.functional as F\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from torch_geometric.nn import GATConv\n",
    "from torch_geometric.data import Data\n",
    "from transformers import AutoTokenizer, AutoModel\n",
    "from tqdm import tqdm\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from typing import List, Tuple, Dict\n",
    "from pykeen.pipeline import pipeline\n",
    "from pykeen.triples import TriplesFactory\n",
    "from torch.utils.data import DataLoader, Dataset\n",
    "from pykeen.training.callbacks import TrainingCallback  # Updated base class for callbacks\n",
    "\n",
    "from pykeen.models import TransR\n",
    "from pykeen.training import SLCWATrainingLoop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import List, Tuple, Dict\n",
    "from torch.utils.data import DataLoader, Dataset\n",
    "\n",
    "class SentenceDataset(Dataset):\n",
    "    def __init__(self, sentences: List[str]):\n",
    "        self.sentences = sentences\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.sentences)\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        return self.sentences[idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "class RNAGPT_LM:\n",
    "    def __init__(self, model_path: str =\"\", model_name: str = \"dna_gpt0.1b_h\", device: str = \"cpu\"):\n",
    "        \"\"\"\n",
    "        RNAGPT-based language model.\n",
    "        Args:\n",
    "            model_name (str): Name of the RNAGPT model. Defaults to 'dna_gpt0.1b_h'.\n",
    "            device (str): Device to run the model on ('cpu' or 'cuda'). Defaults to 'cpu'.\n",
    "        \"\"\"\n",
    "        self.model_path = model_path\n",
    "        self.model_name = model_name\n",
    "        self.device = device\n",
    "        self.model = None\n",
    "        self.tokenizer = None\n",
    "\n",
    "    def get_model(self, model_name):\n",
    "        \"\"\"\n",
    "        Initialize the RNAGPT model and tokenizer.\n",
    "        \"\"\"\n",
    "        special_tokens = (\n",
    "            ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'] +\n",
    "            [\"+\", '-', '*', '/', '=', \"&\", \"|\", \"!\"] +\n",
    "            ['M', 'B'] + ['P'] + ['R', 'I', 'K', 'L', 'O', 'Q', 'S', 'U', 'V'] + ['W', 'Y', 'X', 'Z']\n",
    "        )\n",
    "        if model_name in ('dna_gpt0.1b_h'):\n",
    "            tokenizer = KmerTokenizer(6, special_tokens, dynamic_kmer=False)\n",
    "        else:\n",
    "            tokenizer = KmerTokenizer(6, special_tokens, dynamic_kmer=True)\n",
    "\n",
    "        vocab_size = len(tokenizer)\n",
    "        model = DNAGPT.from_name(model_name, vocab_size)\n",
    "        return model, tokenizer\n",
    "\n",
    "    def load_model(self):\n",
    "        \"\"\"\n",
    "        Load the RNAGPT model and tokenizer from the specified model name.\n",
    "        \"\"\"\n",
    "        self.model, self.tokenizer = self.get_model(self.model_name)\n",
    "        weight_path = os.path.join(self.model_path, f\"{self.model_name}.pth\")\n",
    "        self._load_model_weights(weight_path)\n",
    "        return self\n",
    "\n",
    "    def _load_model_weights(self, weight_path, dtype=None):\n",
    "        \"\"\"\n",
    "        Load the model weights from a checkpoint file.\n",
    "        \"\"\"\n",
    "        state = torch.load(weight_path, map_location=\"cpu\")\n",
    "        if 'model' in state.keys():\n",
    "            self.model.load_state_dict(state['model'], strict=False)\n",
    "        else:\n",
    "            self.model.load_state_dict(state, strict=False)\n",
    "        print(f\"loading model weights from {weight_path}\")\n",
    "        self.model.to(device=self.device, dtype=dtype)\n",
    "        self.model.eval()\n",
    "\n",
    "    def replace_rna_to_dna(self, sequence):\n",
    "        # If the input is a list or tuple, extract the first element.\n",
    "        if isinstance(sequence, (list, tuple)):\n",
    "            sequence = sequence[0]\n",
    "        # If the sequence is a float (or any non-string), return 'N'.\n",
    "        if isinstance(sequence, float):\n",
    "            return 'N'\n",
    "        # Otherwise, replace 'U' with 'T' in the string.\n",
    "        return sequence.replace('U', 'T')\n",
    "    \n",
    "    def generate_embeddings(self, sequences: List[str], batch_size: int = 16, max_len: int = 256, text_emb_dim: int = 64) -> np.ndarray:\n",
    "        \"\"\"\n",
    "        Generate tensor embeddings for each RNA sequence. \n",
    "        Args:\n",
    "            sequences (List[str]): List of RNA sequences.\n",
    "            batch_size (int): Batch size for processing sequences. Defaults to 16.\n",
    "            max_len (int): Maximum length of the sequences. Defaults to 256.\n",
    "            text_emb_dim (int): Dimension to which embeddings are pooled. Defaults to 64.\n",
    "        Returns:\n",
    "            np.ndarray: A numpy array of shape (num_sequences, text_emb_dim)\n",
    "        \"\"\"\n",
    "        print(f\"Generating embeddings for {len(sequences)} sequences.\")\n",
    "        device = self.device\n",
    "        dataset = SentenceDataset(sequences)\n",
    "        # Override collate_fn to return the raw list of sequences\n",
    "        dataloader = DataLoader(dataset, batch_size=batch_size, shuffle=False, collate_fn=lambda x: x)\n",
    "\n",
    "        embeddings = []\n",
    "        for batch in tqdm(dataloader, desc=\"Embedding sequences\", unit=\"sequence\"):\n",
    "            # Tokenize and prepare input tensors as a list of tensors.\n",
    "            batch_tokenized = [\n",
    "                torch.tensor(\n",
    "                    self.tokenizer.encode(\n",
    "                        self.replace_rna_to_dna(seq),\n",
    "                        max_len=max_len,\n",
    "                        device=device\n",
    "                    ),\n",
    "                    dtype=torch.long\n",
    "                )\n",
    "                for seq in batch\n",
    "            ]\n",
    "            # Pad sequences in the batch to the same length\n",
    "            input_ids = torch.nn.utils.rnn.pad_sequence(batch_tokenized, batch_first=True, padding_value=0).to(device)\n",
    "            max_new_tokens = max_len - input_ids.shape[1]\n",
    "\n",
    "            # Forward pass to compute embeddings\n",
    "            with torch.no_grad():\n",
    "                outputs = self.model(input_ids, max_new_tokens)\n",
    "                # Compute the sequence embedding (mean pooling along token dimension)\n",
    "                sequence_embedding = torch.mean(outputs, dim=1).squeeze()\n",
    "                # Adaptive pooling to the desired dimension (pooling over the hidden dimension)\n",
    "                projected = torch.nn.functional.adaptive_avg_pool1d(sequence_embedding.unsqueeze(1), output_size=text_emb_dim).squeeze(1)\n",
    "                embeddings.append(projected)\n",
    "\n",
    "        # Concatenate embeddings from all batches along the first dimension\n",
    "        embeddings_tensor = torch.cat(embeddings, dim=0)\n",
    "        print(f\"Generated embeddings shape: {embeddings_tensor.shape}\")\n",
    "\n",
    "        # Convert to numpy array and return\n",
    "        return embeddings_tensor.cpu().numpy()\n",
    "\n",
    "    def generate_mid_embeddings(self, sequences: List[str], max_len: int = 256) -> torch.Tensor:\n",
    "        \"\"\"\n",
    "        Generate tensor mid embeddings for each RNA sequence. \n",
    "        Args:\n",
    "            sequences (List[str]): List of DNA sequences.\n",
    "            max_len (int): Maximum length of the sequences. Defaults to 256.\n",
    "        Returns:\n",
    "            torch.Tensor: A tensor of shape (num_sequences, embedding_dim) where\n",
    "                        num_sequences is the number of sequences and\n",
    "                        embedding_dim is the dimension of the reduced embeddings.\n",
    "        \"\"\"\n",
    "        print(f\"Generating embeddings for {len(sequences)} sequences.\")\n",
    "        device = self.device\n",
    "        embeddings = []\n",
    "\n",
    "        for sequence in tqdm(sequences, desc=\"Embedding sequences\", unit=\"sequence\"):\n",
    "            # Tokenize and prepare input tensors\n",
    "            sequence = self.replace_rna_to_dna(sequence[0])\n",
    "            input_ids = self.tokenizer.encode(sequence, max_len=max_len, device=device)[None, :]  # Add batch dimension\n",
    "            input_ids = torch.tensor(input_ids, dtype=torch.long, device=device)\n",
    "            max_new_tokens = max_len - input_ids.shape[1]\n",
    "\n",
    "            # Forward pass to compute embeddings\n",
    "            with torch.no_grad():\n",
    "                outputs = self.model.generate_mid_embeddings(input_ids, max_new_tokens)\n",
    "                # Compute the sequence embedding (mean pooling along token dimension)\n",
    "                sequence_embedding = torch.mean(outputs, dim=1).squeeze()\n",
    "                embeddings.append(sequence_embedding)\n",
    "\n",
    "        # Stack embeddings into a single tensor (shape: [num_sequences, embedding_dim])\n",
    "        embeddings_tensor = torch.stack(embeddings)\n",
    "        print(f\"Generated embeddings shape: {embeddings_tensor.shape}\")\n",
    "        return embeddings_tensor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of parameters: 100.13M\n",
      "loading model weights from .\\dna_gpt0.1b_h.pth\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<__main__.RNAGPT_LM at 0x1bccffd2860>"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "device = \"cuda\"\n",
    "rna_gpt_lm = RNAGPT_LM(model_path=\".\", model_name=\"dna_gpt0.1b_h\", device=device)\n",
    "rna_gpt_lm.load_model()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CTGCTGCTGCTGCGCCCCATCCCCCCGCGGCCGGCCAGTTCCAGCCCGCACCCCGCGTCGGTGCCCGCGCCCCTCCCCGGGCCCCGCCATGGGCCTCACCGTGTCCGCGCTCTTTTCGCGGATCTTCGGGAAGAAGCAGATGCGGATTCTCATGGGTGAGGCAGATCGAGCGCGCGGCCCGGACCGGGGCGCCGGCCCCGGCGCAGCCCTTCCGCCCCCGCGTCCCTCCAGCCCCGCTCACCTGGGTCTCTGGCCCCGAGTCACCCACCTCATAACGCCCCGGGGCCTCTGCTCTCGGGCGGGTCCCGGTCTGCATCGCCGACCCCGGGGCCTGACACCCGGAGCTGCGGGCCTGGGTGGGGTGAAGCTCCCTCCGCCCCAGCCCGGCTGGTAAGAAGGGAGGATTCCGCCCTTGGAGACGACTTTTAAAACGAGCGCGGCCTACCTCCCGTGCCCCTTGCCTCCCAGTCCTCTCCCGCTCCGCGCCCTCTTTGGAGTTGGCTCACGCCACCCGACTGCCCTCCAGGCCTCAAAGGTAGATAACGACGCGCACCTGGAGGCGCTCCCGCTCTCCGCCCCAGTCACCATCAGCTGTTGTGGCCTCTCCCTTTCCCTGGTCTCTAGGGGGCTCCCTCGCTCCCATCTCCATCCCTGTGCCCCTTTCCGTTGCAGTTGGCTTGGATGCGGCTGGCAAGACCACAATCCTGTACAAACTGAAGTTGGGGGAGATTGTCACCACCATCCCAACCATAGGTGAGCCCGGGGACGAAGCAGGGAGCGGGAGCGCCGCGGCGGGCCCTCGCGGGGTCTGCTCCCAGTTCTGCCACTTTTCTGGAGCTGACCCTGACACCAGATACAGCTATTTGAGAAGTGGGTCAGGGTATCTCTACGTGGATACCGGAGGCAGGGTGGATGTGACCTAGCCCCGCCCTGTTGACCGCCAGTTCTGGGGTTCTGTTCCCCTGGGCCTTGATCATTGCCTTCTGGTTTTGTGTCCCTGCTGAAATCTAAGTAGTTTTAGTGAGTTCCTTTCTTTCAGGAGTTTTCTCATCTTTTTTTTCCAATTATCTGCAGGCTTCAATGTAGAAACAGTGGAATATAAGAACATCTGTTTCACAGTCTGGGACGTGGGAGGCCAGGACAAGATTCGGCCTCTGTGGCGGCACTACTTCCAGAACACTCAGGTGGAGTGTTGGGAGGGGACTTTCTAACCCCACGGGAAAAGGTGTTAGGGCTGGGAGAACAGAATTGTTGGCCTAGGCTGGGCCCCCAGGCCAAGGAAAAAACCCTTCCCCACTTCTACCCCTTCTGTTTGGCTATCCCCTACTCACTCTTCAGAACTCACCTTAGTCTTCCCCAGCTTGGACTTCTCTTTAGAAGCAGGATTCTGGGACCAGGCTGCCTGATTTAGCCCTCATAACATGTCTTTATTTCCTTGCACATCTTTGGATAAACTACTTCCTCTTTGCAGTGGGGTCTCTAGTTTGTGCCATGACTCCTGTAGAGCTTCTATAGAGTTCTGTAAACTTCTGTAGAGTTTACTAGATGAGAAGGATTGACACAATTACAAATGGGAAGAAAACAGAAAGGGCCACACTACGGTATGTCCCAGGCAGAAGTGATTGCTTCTCCTTTCTTCCTTCCTTCCTGCCCAGGGCCTCATCTTTGTGGTGGACAGTAATGACCGGGAGCGGGTCCAAGAATCTGCTGATGAACTCCAGAAGATGGTGAGTACCCAGAGCCCTGGGAACTGAGCCCTCAGCTTGGGGACAGAGTGATCTCTGTAGTGGTATAGAAGTCAGGGAGCCCCCAACAGGCATTGAAGACCAGGAATATAAGTTTTTCTTTGTGGACAGACACATTGTGTATCTCACCCTGTTTTGGATGGGAGTGACTTTTTCACTTTTTCTGTGCATTGTCTTGTCTTTTTTTTTTTTTGAAATGGGAGTTTTGCTGTTGTTGCCCAGGCTGGAGTGCAATGGTGCAATCTCGGCTCACTGCAACCTCCGTCTCCCGGGTTCAAGCAATTCTCCTGCCTCAGCCTCCCAAGTAGCTGGGATTACAGGCGCCCGCTGCCATGCCCAGTTAATTTTTGTATTTTTAGTAGAGACGGGGTTTCACCATGTTGGCCAGGCTGGTCTTGAACCCCCAACCTCAGGTGATCCACCCGTCTTGGTCTCCCAAAGTGCTGGGATTACAGGAGTGAGCCACTGCGCCCGGCCGCATTCTCATGTCTTTACATCTGAGTCTGAGAAGTTGTGGGAGAGATGGGACTTATATGTGGCAAAGAGAAACATTTTACATTCCTGCCTTTTCCATTTGGAAATGTTGTCATTTTGCCTAGGCAAGTACTCAAATTAGGAAAAGGTCTTTCTTTTCTGGACTTATGGGCTCAGGTCACCTCAAGCTCATTTACAACTATGTCTTGTCTTCAACGGCAGACTGCACAATACTTTTTTTTTCCAATCAAGATGCTAGGAGGTAGAAAGGGAATTCCCCAACACAGTAGAGGAGACGCAGCCTGGGTCCCACCTTCTCTTCTCCTCTCTCAGCTGCAGGAGGACGAGCTGCGGGATGCAGTGCTGCTGGTATTTGCCAACAAGCAGGACATGCCCAACGCCATGCCCGTGAGCGAGCTGACTGACAAGCTGGGGCTACAGCACTTACGCAGCCGCACGGTAGGGGTCCTGCCCACCTGGTGCTGAATCCTGCCTCTTGAGGGAAGCTGCAGGCTGGGACAGAGATATAAAGGCATTCCTTTTGTTCCTGGTTGCTGACCTCTTTCTTTCCTTTTCCCCACAGTGGTATGTCCAGGCCACCTGTGCCACCCAAGGCACAGGTCTGTACGATGGTCTGGACTGGCTGTCCCACGAGCTGTCAAAGCGCTAACCAGCCAGGGGCAGGCCCCTGATGCCCGGAAGCTCCTGCGTGCATCCCCGGGATGACCAGACTCCCGGACTCCTCAGGCAGTGCCCTTTCCTCCCACTTTTCCTCCCCCATAGCCACAGGCCTCTGCTCCTGCTCCTGCCTGCATGTTCTCTCTGTTGTTGGAGCCTGGAGCCTTGCTCTCTGGGCACAGAGGGGTCCACTCTCCTGCCTGCTGGGACCTATGGAAGGGGCTTCCTGGCCAAGGCCCCCTCTTCCAGAGGAGGAGCAGGGATCTGGGTTTCCTTTTTTTTTTCTGTTTTGGGTGTACTCTAGGGGCCAGGTTGGGAGGGGGAAGGTGAGGGCTTCGGGTGGTGCTATAATGTGGCACTGGATCTTGAGTAATAAATTTGCTGTGGTTTGTA'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_sequences[0].replace('U', 'T')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "278326"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(transcript_sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace RNA to DNA by \"sequence.replace('U', 'T')\" to the list of sequences\n",
    "replaced_transcript_sequences = []\n",
    "for i in transcript_sequences:\n",
    "    if type(i) == float:\n",
    "        replaced_transcript_sequences.append('')\n",
    "    else:\n",
    "        replaced_transcript_sequences.append(i.replace('U', 'T'))\n",
    " \n",
    "len(replaced_transcript_sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating embeddings for 278326 sequences.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Embedding sequences:   0%|          | 0/4349 [00:00<?, ?sequence/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\hemingzhang\\AppData\\Local\\Temp\\ipykernel_25784\\2961308162.py:86: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).\n",
      "  torch.tensor(\n",
      "Embedding sequences: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 4349/4349 [1:01:16<00:00,  1.18sequence/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generated embeddings shape: torch.Size([278326, 1])\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "rna_seq = transcript_sequences\n",
    "batch_size = 64\n",
    "rna_embedding = rna_gpt_lm.generate_embeddings(sequences=rna_seq, batch_size=batch_size, max_len=256, text_emb_dim=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_npy_path = \"./CellTOSG/x_ts_emb.npy\"\n",
    "np.save(output_npy_path, rna_embedding)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Merge Protein Sequence to BMG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>contig</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>dna_sequence</th>\n",
       "      <th>transcript_sequence</th>\n",
       "      <th>protein_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000147655</td>\n",
       "      <td>RSPO2</td>\n",
       "      <td>8</td>\n",
       "      <td>107899316.0</td>\n",
       "      <td>108083642.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>AGAAGCATCCCAAGCCTCAGCCGGTCTGCATCTCCATCGGAAAGTG...</td>\n",
       "      <td>GGUCUGCAUCUCCAUCGGAAAGUGCGCUUGCCACAUCCCUUCGGAU...</td>\n",
       "      <td>MRQYGECLHSCPSGYYGHRAPDMNRCARCRIENCDSCFSKDFCTKC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000147669</td>\n",
       "      <td>POLR2K</td>\n",
       "      <td>8</td>\n",
       "      <td>100150623.0</td>\n",
       "      <td>100154003.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ATGTACCAGCGCCGGAAGTTGGTCTCGACACCTGGACTAGCCGGGT...</td>\n",
       "      <td>GGAAGUUGGUCUCGACACCUGGACUAGCCGGGUUGUAUUUGGAAAC...</td>\n",
       "      <td>MDTQKDVQPPKQQPMIYICGECHTENEIKSRDPIRCRECGYRIMYK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000147676</td>\n",
       "      <td>MAL2</td>\n",
       "      <td>8</td>\n",
       "      <td>119165034.0</td>\n",
       "      <td>119245673.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>AGATTCCATGGGCGTAGGACCCTCCGAGCCAGGTGTGGGATATAAT...</td>\n",
       "      <td>GCGCCCGGAGCCCGCGGAGCUGAGCGGCGGCGGCGGCGGCGGCAGG...</td>\n",
       "      <td>MFVSVTAFFFSLLFLGMFLSGMVAQIDANWNFLDFAYHFTVFVFYF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000147677</td>\n",
       "      <td>EIF3H</td>\n",
       "      <td>8</td>\n",
       "      <td>116642130.0</td>\n",
       "      <td>116766925.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>ACTTAACTGTCAACTCCACGTGGGCGGGGCGTTTTGCACCTCCTGA...</td>\n",
       "      <td>CUCUUUCUUCCUGUCUGCUUGGAAAGAUGGCGUCCCGCAAGGAAGG...</td>\n",
       "      <td>MASRKEGTGSTATSSSSTAGAAGKGKGKGGSGDSAVKQVQIDGLVV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000147679</td>\n",
       "      <td>UTP23</td>\n",
       "      <td>8</td>\n",
       "      <td>116766505.0</td>\n",
       "      <td>116849463.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TTCATTTCCGGGTGAAACTGGCATTGAGGGTACTGGGGCGTGCGTG...</td>\n",
       "      <td>GGCAUUGAGGGUACUGGGGCGUGCGUGAGGCGUUUACUGAUGCUUC...</td>\n",
       "      <td>MKITRQKHAKKHLGFFRNNFGVREPYQILLDGTFCQAALRGRIQLR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121401</th>\n",
       "      <td>ENSG00000147642</td>\n",
       "      <td>SYBU</td>\n",
       "      <td>8</td>\n",
       "      <td>109573978.0</td>\n",
       "      <td>109691791.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>AGCGGCTTCTTCAAGAAACAGATCCTGCAGCATAATCTGCCAGGGC...</td>\n",
       "      <td>AGAGCAAUCGGCGGCAGUGGCGGCGGCGGCGCGUUGGAGCCUCACC...</td>\n",
       "      <td>MGPLRESKKEHRVQHHDKEISRSRIPRLILRPHMPQQQHKVSPASE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121402</th>\n",
       "      <td>ENSG00000147647</td>\n",
       "      <td>DPYS</td>\n",
       "      <td>8</td>\n",
       "      <td>104330324.0</td>\n",
       "      <td>104467055.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>GGATTTGCAGCCTGAGGCAGAGCTCGGGGGCTGTCGGTGGGGACCT...</td>\n",
       "      <td>GGAUUUGCAGCCUGAGGCAGAGCUCGGGGGCUGUCGGUGGGGACCU...</td>\n",
       "      <td>MAAPSRLLIRGGRVVNDDFSEVADVLVEDGVVRALGHDLLPPGGAP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121403</th>\n",
       "      <td>ENSG00000147649</td>\n",
       "      <td>MTDH</td>\n",
       "      <td>8</td>\n",
       "      <td>97644184.0</td>\n",
       "      <td>97730260.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>AAGCCGCCATTGTTCCGCCGAGGGAGGACAGCGGGGCCTGGCGCTG...</td>\n",
       "      <td>AAGCCGCCAUUGUUCCGCCGAGGGAGGACAGCGGGGCCUGGCGCUG...</td>\n",
       "      <td>MAARSWQDELAQQAEEGSARLREMLSVGLGFLRTELGLDLGLEPKR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121404</th>\n",
       "      <td>ENSG00000147650</td>\n",
       "      <td>LRP12</td>\n",
       "      <td>8</td>\n",
       "      <td>104489231.0</td>\n",
       "      <td>104589258.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>AGTCTGGCGAGTGGCGGCGCCGGCTAGCCCGAGCGGCGCGAGGCCG...</td>\n",
       "      <td>AGUCUGGCGAGUGGCGGCGCCGGCUAGCCCGAGCGGCGCGAGGCCG...</td>\n",
       "      <td>MACRWSTKESPRWRSALLLLFLAGVYACGETPEQIRAPSGIITSPG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121405</th>\n",
       "      <td>ENSG00000147654</td>\n",
       "      <td>EBAG9</td>\n",
       "      <td>8</td>\n",
       "      <td>109539711.0</td>\n",
       "      <td>109570072.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GATGAAGGGGCGGCCATGGCAGCTGCGCAGAGGCAACGCAGGCTGC...</td>\n",
       "      <td>AGGCGUGCAGCAUUCGCCAUGCUCCGCUCACGCGUGGGAGACUGGG...</td>\n",
       "      <td>MAITQFRLFKFCTCLATVFSFLKRLICRSGRGRKLSGDQITLPTTV...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121406 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_id gene_name contig        start          end  strand  \\\n",
       "0       ENSG00000147655     RSPO2      8  107899316.0  108083642.0    -1.0   \n",
       "1       ENSG00000147669    POLR2K      8  100150623.0  100154003.0     1.0   \n",
       "2       ENSG00000147676      MAL2      8  119165034.0  119245673.0     1.0   \n",
       "3       ENSG00000147677     EIF3H      8  116642130.0  116766925.0    -1.0   \n",
       "4       ENSG00000147679     UTP23      8  116766505.0  116849463.0     1.0   \n",
       "...                 ...       ...    ...          ...          ...     ...   \n",
       "121401  ENSG00000147642      SYBU      8  109573978.0  109691791.0    -1.0   \n",
       "121402  ENSG00000147647      DPYS      8  104330324.0  104467055.0    -1.0   \n",
       "121403  ENSG00000147649      MTDH      8   97644184.0   97730260.0     1.0   \n",
       "121404  ENSG00000147650     LRP12      8  104489231.0  104589258.0    -1.0   \n",
       "121405  ENSG00000147654     EBAG9      8  109539711.0  109570072.0     1.0   \n",
       "\n",
       "                                             dna_sequence  \\\n",
       "0       AGAAGCATCCCAAGCCTCAGCCGGTCTGCATCTCCATCGGAAAGTG...   \n",
       "1       ATGTACCAGCGCCGGAAGTTGGTCTCGACACCTGGACTAGCCGGGT...   \n",
       "2       AGATTCCATGGGCGTAGGACCCTCCGAGCCAGGTGTGGGATATAAT...   \n",
       "3       ACTTAACTGTCAACTCCACGTGGGCGGGGCGTTTTGCACCTCCTGA...   \n",
       "4       TTCATTTCCGGGTGAAACTGGCATTGAGGGTACTGGGGCGTGCGTG...   \n",
       "...                                                   ...   \n",
       "121401  AGCGGCTTCTTCAAGAAACAGATCCTGCAGCATAATCTGCCAGGGC...   \n",
       "121402  GGATTTGCAGCCTGAGGCAGAGCTCGGGGGCTGTCGGTGGGGACCT...   \n",
       "121403  AAGCCGCCATTGTTCCGCCGAGGGAGGACAGCGGGGCCTGGCGCTG...   \n",
       "121404  AGTCTGGCGAGTGGCGGCGCCGGCTAGCCCGAGCGGCGCGAGGCCG...   \n",
       "121405  GATGAAGGGGCGGCCATGGCAGCTGCGCAGAGGCAACGCAGGCTGC...   \n",
       "\n",
       "                                      transcript_sequence  \\\n",
       "0       GGUCUGCAUCUCCAUCGGAAAGUGCGCUUGCCACAUCCCUUCGGAU...   \n",
       "1       GGAAGUUGGUCUCGACACCUGGACUAGCCGGGUUGUAUUUGGAAAC...   \n",
       "2       GCGCCCGGAGCCCGCGGAGCUGAGCGGCGGCGGCGGCGGCGGCAGG...   \n",
       "3       CUCUUUCUUCCUGUCUGCUUGGAAAGAUGGCGUCCCGCAAGGAAGG...   \n",
       "4       GGCAUUGAGGGUACUGGGGCGUGCGUGAGGCGUUUACUGAUGCUUC...   \n",
       "...                                                   ...   \n",
       "121401  AGAGCAAUCGGCGGCAGUGGCGGCGGCGGCGCGUUGGAGCCUCACC...   \n",
       "121402  GGAUUUGCAGCCUGAGGCAGAGCUCGGGGGCUGUCGGUGGGGACCU...   \n",
       "121403  AAGCCGCCAUUGUUCCGCCGAGGGAGGACAGCGGGGCCUGGCGCUG...   \n",
       "121404  AGUCUGGCGAGUGGCGGCGCCGGCUAGCCCGAGCGGCGCGAGGCCG...   \n",
       "121405  AGGCGUGCAGCAUUCGCCAUGCUCCGCUCACGCGUGGGAGACUGGG...   \n",
       "\n",
       "                                         protein_sequence  \n",
       "0       MRQYGECLHSCPSGYYGHRAPDMNRCARCRIENCDSCFSKDFCTKC...  \n",
       "1       MDTQKDVQPPKQQPMIYICGECHTENEIKSRDPIRCRECGYRIMYK...  \n",
       "2       MFVSVTAFFFSLLFLGMFLSGMVAQIDANWNFLDFAYHFTVFVFYF...  \n",
       "3       MASRKEGTGSTATSSSSTAGAAGKGKGKGGSGDSAVKQVQIDGLVV...  \n",
       "4       MKITRQKHAKKHLGFFRNNFGVREPYQILLDGTFCQAALRGRIQLR...  \n",
       "...                                                   ...  \n",
       "121401  MGPLRESKKEHRVQHHDKEISRSRIPRLILRPHMPQQQHKVSPASE...  \n",
       "121402  MAAPSRLLIRGGRVVNDDFSEVADVLVEDGVVRALGHDLLPPGGAP...  \n",
       "121403  MAARSWQDELAQQAEEGSARLREMLSVGLGFLRTELGLDLGLEPKR...  \n",
       "121404  MACRWSTKESPRWRSALLLLFLAGVYACGETPEQIRAPSGIITSPG...  \n",
       "121405  MAITQFRLFKFCTCLATVFSFLKRLICRSGRGRKLSGDQITLPTTV...  \n",
       "\n",
       "[121406 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sequence_df = pd.read_csv('./BioMedGraphica/combined_sequences.csv')\n",
    "display(sequence_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>protein_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000147655</td>\n",
       "      <td>RSPO2</td>\n",
       "      <td>MRQYGECLHSCPSGYYGHRAPDMNRCARCRIENCDSCFSKDFCTKC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000147669</td>\n",
       "      <td>POLR2K</td>\n",
       "      <td>MDTQKDVQPPKQQPMIYICGECHTENEIKSRDPIRCRECGYRIMYK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000147676</td>\n",
       "      <td>MAL2</td>\n",
       "      <td>MFVSVTAFFFSLLFLGMFLSGMVAQIDANWNFLDFAYHFTVFVFYF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000147677</td>\n",
       "      <td>EIF3H</td>\n",
       "      <td>MASRKEGTGSTATSSSSTAGAAGKGKGKGGSGDSAVKQVQIDGLVV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000147679</td>\n",
       "      <td>UTP23</td>\n",
       "      <td>MKITRQKHAKKHLGFFRNNFGVREPYQILLDGTFCQAALRGRIQLR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121401</th>\n",
       "      <td>ENSG00000147642</td>\n",
       "      <td>SYBU</td>\n",
       "      <td>MGPLRESKKEHRVQHHDKEISRSRIPRLILRPHMPQQQHKVSPASE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121402</th>\n",
       "      <td>ENSG00000147647</td>\n",
       "      <td>DPYS</td>\n",
       "      <td>MAAPSRLLIRGGRVVNDDFSEVADVLVEDGVVRALGHDLLPPGGAP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121403</th>\n",
       "      <td>ENSG00000147649</td>\n",
       "      <td>MTDH</td>\n",
       "      <td>MAARSWQDELAQQAEEGSARLREMLSVGLGFLRTELGLDLGLEPKR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121404</th>\n",
       "      <td>ENSG00000147650</td>\n",
       "      <td>LRP12</td>\n",
       "      <td>MACRWSTKESPRWRSALLLLFLAGVYACGETPEQIRAPSGIITSPG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121405</th>\n",
       "      <td>ENSG00000147654</td>\n",
       "      <td>EBAG9</td>\n",
       "      <td>MAITQFRLFKFCTCLATVFSFLKRLICRSGRGRKLSGDQITLPTTV...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121406 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_id gene_name  \\\n",
       "0       ENSG00000147655     RSPO2   \n",
       "1       ENSG00000147669    POLR2K   \n",
       "2       ENSG00000147676      MAL2   \n",
       "3       ENSG00000147677     EIF3H   \n",
       "4       ENSG00000147679     UTP23   \n",
       "...                 ...       ...   \n",
       "121401  ENSG00000147642      SYBU   \n",
       "121402  ENSG00000147647      DPYS   \n",
       "121403  ENSG00000147649      MTDH   \n",
       "121404  ENSG00000147650     LRP12   \n",
       "121405  ENSG00000147654     EBAG9   \n",
       "\n",
       "                                         protein_sequence  \n",
       "0       MRQYGECLHSCPSGYYGHRAPDMNRCARCRIENCDSCFSKDFCTKC...  \n",
       "1       MDTQKDVQPPKQQPMIYICGECHTENEIKSRDPIRCRECGYRIMYK...  \n",
       "2       MFVSVTAFFFSLLFLGMFLSGMVAQIDANWNFLDFAYHFTVFVFYF...  \n",
       "3       MASRKEGTGSTATSSSSTAGAAGKGKGKGGSGDSAVKQVQIDGLVV...  \n",
       "4       MKITRQKHAKKHLGFFRNNFGVREPYQILLDGTFCQAALRGRIQLR...  \n",
       "...                                                   ...  \n",
       "121401  MGPLRESKKEHRVQHHDKEISRSRIPRLILRPHMPQQQHKVSPASE...  \n",
       "121402  MAAPSRLLIRGGRVVNDDFSEVADVLVEDGVVRALGHDLLPPGGAP...  \n",
       "121403  MAARSWQDELAQQAEEGSARLREMLSVGLGFLRTELGLDLGLEPKR...  \n",
       "121404  MACRWSTKESPRWRSALLLLFLAGVYACGETPEQIRAPSGIITSPG...  \n",
       "121405  MAITQFRLFKFCTCLATVFSFLKRLICRSGRGRKLSGDQITLPTTV...  \n",
       "\n",
       "[121406 rows x 3 columns]"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_seq_df = sequence_df[['gene_id', 'gene_name', 'protein_sequence']]\n",
    "protein_seq_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\hemingzhang\\AppData\\Local\\Temp\\ipykernel_25784\\1568805577.py:1: DtypeWarning: Columns (1,2,5,7) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  protein_df = pd.read_csv('./BioMedGraphica/Entity/Protein/biomedgraphica_protein.csv')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioMedGraphica_ID</th>\n",
       "      <th>Ensembl_Protein_ID</th>\n",
       "      <th>Ensembl_Protein_ID_Version</th>\n",
       "      <th>Uniprot_ID</th>\n",
       "      <th>RefSeq_ID</th>\n",
       "      <th>NCBI_ID</th>\n",
       "      <th>Protein_Name</th>\n",
       "      <th>HGNC_Symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BMG_PT000001</td>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000000233.5</td>\n",
       "      <td>P84085</td>\n",
       "      <td>NP_001653;NP_001653.1</td>\n",
       "      <td>381</td>\n",
       "      <td>ADP-ribosylation factor 5</td>\n",
       "      <td>ARF5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BMG_PT000002</td>\n",
       "      <td>ENSP00000000412</td>\n",
       "      <td>ENSP00000000412.3</td>\n",
       "      <td>P20645</td>\n",
       "      <td>NP_001401249;NP_001401249.1;NP_001401260;NP_00...</td>\n",
       "      <td>4074</td>\n",
       "      <td>Cation-dependent mannose-6-phosphate receptor ...</td>\n",
       "      <td>M6PR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BMG_PT000003</td>\n",
       "      <td>ENSP00000000442</td>\n",
       "      <td>ENSP00000000442.6</td>\n",
       "      <td>P11474</td>\n",
       "      <td>NP_001269379.1;NP_001269380;NP_004442;NP_004442.3</td>\n",
       "      <td>2101</td>\n",
       "      <td>Steroid hormone receptor ERR1 (Estrogen recept...</td>\n",
       "      <td>ESRRA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BMG_PT000004</td>\n",
       "      <td>ENSP00000001008</td>\n",
       "      <td>ENSP00000001008.4</td>\n",
       "      <td>Q02790</td>\n",
       "      <td>NP_002005;NP_002005.1</td>\n",
       "      <td>2288</td>\n",
       "      <td>Peptidyl-prolyl cis-trans isomerase FKBP4 (PPI...</td>\n",
       "      <td>FKBP4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BMG_PT000005</td>\n",
       "      <td>ENSP00000001146</td>\n",
       "      <td>ENSP00000001146.2</td>\n",
       "      <td>Q9NR63</td>\n",
       "      <td>NP_001264671;NP_063938;NP_063938.1</td>\n",
       "      <td>56603</td>\n",
       "      <td>Cytochrome P450 26B1 (EC 1.14.13.-) (Cytochrom...</td>\n",
       "      <td>CYP26B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205486</th>\n",
       "      <td>BMG_PT205487</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X6RGR3</td>\n",
       "      <td>NP_001353609.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205487</th>\n",
       "      <td>BMG_PT205488</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X6RLR1</td>\n",
       "      <td>NP_001268356.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205488</th>\n",
       "      <td>BMG_PT205489</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X6RLX0</td>\n",
       "      <td>XP_054227478.1;XP_016874551.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205489</th>\n",
       "      <td>BMG_PT205490</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X6RM00</td>\n",
       "      <td>XP_047284540.1;XP_054227509.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205490</th>\n",
       "      <td>BMG_PT205491</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>X6RM59</td>\n",
       "      <td>NP_001002010.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>205491 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       BioMedGraphica_ID Ensembl_Protein_ID Ensembl_Protein_ID_Version  \\\n",
       "0           BMG_PT000001    ENSP00000000233          ENSP00000000233.5   \n",
       "1           BMG_PT000002    ENSP00000000412          ENSP00000000412.3   \n",
       "2           BMG_PT000003    ENSP00000000442          ENSP00000000442.6   \n",
       "3           BMG_PT000004    ENSP00000001008          ENSP00000001008.4   \n",
       "4           BMG_PT000005    ENSP00000001146          ENSP00000001146.2   \n",
       "...                  ...                ...                        ...   \n",
       "205486      BMG_PT205487                NaN                        NaN   \n",
       "205487      BMG_PT205488                NaN                        NaN   \n",
       "205488      BMG_PT205489                NaN                        NaN   \n",
       "205489      BMG_PT205490                NaN                        NaN   \n",
       "205490      BMG_PT205491                NaN                        NaN   \n",
       "\n",
       "       Uniprot_ID                                          RefSeq_ID NCBI_ID  \\\n",
       "0          P84085                              NP_001653;NP_001653.1     381   \n",
       "1          P20645  NP_001401249;NP_001401249.1;NP_001401260;NP_00...    4074   \n",
       "2          P11474  NP_001269379.1;NP_001269380;NP_004442;NP_004442.3    2101   \n",
       "3          Q02790                              NP_002005;NP_002005.1    2288   \n",
       "4          Q9NR63                 NP_001264671;NP_063938;NP_063938.1   56603   \n",
       "...           ...                                                ...     ...   \n",
       "205486     X6RGR3                                     NP_001353609.1     NaN   \n",
       "205487     X6RLR1                                     NP_001268356.1     NaN   \n",
       "205488     X6RLX0                      XP_054227478.1;XP_016874551.1     NaN   \n",
       "205489     X6RM00                      XP_047284540.1;XP_054227509.1     NaN   \n",
       "205490     X6RM59                                     NP_001002010.2     NaN   \n",
       "\n",
       "                                             Protein_Name HGNC_Symbol  \n",
       "0                               ADP-ribosylation factor 5        ARF5  \n",
       "1       Cation-dependent mannose-6-phosphate receptor ...        M6PR  \n",
       "2       Steroid hormone receptor ERR1 (Estrogen recept...       ESRRA  \n",
       "3       Peptidyl-prolyl cis-trans isomerase FKBP4 (PPI...       FKBP4  \n",
       "4       Cytochrome P450 26B1 (EC 1.14.13.-) (Cytochrom...     CYP26B1  \n",
       "...                                                   ...         ...  \n",
       "205486                                                NaN         NaN  \n",
       "205487                                                NaN         NaN  \n",
       "205488                                                NaN         NaN  \n",
       "205489                                                NaN         NaN  \n",
       "205490                                                NaN         NaN  \n",
       "\n",
       "[205491 rows x 8 columns]"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_df = pd.read_csv('./BioMedGraphica/Entity/Protein/biomedgraphica_protein.csv')\n",
    "protein_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioMedGraphica_ID</th>\n",
       "      <th>HGNC_Symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BMG_PT000001</td>\n",
       "      <td>ARF5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BMG_PT000002</td>\n",
       "      <td>M6PR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BMG_PT000003</td>\n",
       "      <td>ESRRA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BMG_PT000004</td>\n",
       "      <td>FKBP4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BMG_PT000005</td>\n",
       "      <td>CYP26B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205486</th>\n",
       "      <td>BMG_PT205487</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205487</th>\n",
       "      <td>BMG_PT205488</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205488</th>\n",
       "      <td>BMG_PT205489</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205489</th>\n",
       "      <td>BMG_PT205490</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205490</th>\n",
       "      <td>BMG_PT205491</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>205491 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       BioMedGraphica_ID HGNC_Symbol\n",
       "0           BMG_PT000001        ARF5\n",
       "1           BMG_PT000002        M6PR\n",
       "2           BMG_PT000003       ESRRA\n",
       "3           BMG_PT000004       FKBP4\n",
       "4           BMG_PT000005     CYP26B1\n",
       "...                  ...         ...\n",
       "205486      BMG_PT205487         NaN\n",
       "205487      BMG_PT205488         NaN\n",
       "205488      BMG_PT205489         NaN\n",
       "205489      BMG_PT205490         NaN\n",
       "205490      BMG_PT205491         NaN\n",
       "\n",
       "[205491 rows x 2 columns]"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_id_df = protein_df[['BioMedGraphica_ID', 'HGNC_Symbol']]\n",
    "protein_id_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>protein_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RSPO2</td>\n",
       "      <td>MRQYGECLHSCPSGYYGHRAPDMNRCARCRIENCDSCFSKDFCTKC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>POLR2K</td>\n",
       "      <td>MDTQKDVQPPKQQPMIYICGECHTENEIKSRDPIRCRECGYRIMYK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MAL2</td>\n",
       "      <td>MFVSVTAFFFSLLFLGMFLSGMVAQIDANWNFLDFAYHFTVFVFYF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EIF3H</td>\n",
       "      <td>MASRKEGTGSTATSSSSTAGAAGKGKGKGGSGDSAVKQVQIDGLVV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UTP23</td>\n",
       "      <td>MKITRQKHAKKHLGFFRNNFGVREPYQILLDGTFCQAALRGRIQLR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17822</th>\n",
       "      <td>SYBU</td>\n",
       "      <td>MGPLRESKKEHRVQHHDKEISRSRIPRLILRPHMPQQQHKVSPASE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17823</th>\n",
       "      <td>DPYS</td>\n",
       "      <td>MAAPSRLLIRGGRVVNDDFSEVADVLVEDGVVRALGHDLLPPGGAP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17824</th>\n",
       "      <td>MTDH</td>\n",
       "      <td>MAARSWQDELAQQAEEGSARLREMLSVGLGFLRTELGLDLGLEPKR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17825</th>\n",
       "      <td>LRP12</td>\n",
       "      <td>MACRWSTKESPRWRSALLLLFLAGVYACGETPEQIRAPSGIITSPG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17826</th>\n",
       "      <td>EBAG9</td>\n",
       "      <td>MAITQFRLFKFCTCLATVFSFLKRLICRSGRGRKLSGDQITLPTTV...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17827 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name                                   protein_sequence\n",
       "0         RSPO2  MRQYGECLHSCPSGYYGHRAPDMNRCARCRIENCDSCFSKDFCTKC...\n",
       "1        POLR2K  MDTQKDVQPPKQQPMIYICGECHTENEIKSRDPIRCRECGYRIMYK...\n",
       "2          MAL2  MFVSVTAFFFSLLFLGMFLSGMVAQIDANWNFLDFAYHFTVFVFYF...\n",
       "3         EIF3H  MASRKEGTGSTATSSSSTAGAAGKGKGKGGSGDSAVKQVQIDGLVV...\n",
       "4         UTP23  MKITRQKHAKKHLGFFRNNFGVREPYQILLDGTFCQAALRGRIQLR...\n",
       "...         ...                                                ...\n",
       "17822      SYBU  MGPLRESKKEHRVQHHDKEISRSRIPRLILRPHMPQQQHKVSPASE...\n",
       "17823      DPYS  MAAPSRLLIRGGRVVNDDFSEVADVLVEDGVVRALGHDLLPPGGAP...\n",
       "17824      MTDH  MAARSWQDELAQQAEEGSARLREMLSVGLGFLRTELGLDLGLEPKR...\n",
       "17825     LRP12  MACRWSTKESPRWRSALLLLFLAGVYACGETPEQIRAPSGIITSPG...\n",
       "17826     EBAG9  MAITQFRLFKFCTCLATVFSFLKRLICRSGRGRKLSGDQITLPTTV...\n",
       "\n",
       "[17827 rows x 2 columns]"
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_seq_dropna_df = protein_seq_df.drop_duplicates(subset='gene_name').dropna().reset_index(drop=True)[['gene_name', 'protein_sequence']]\n",
    "protein_seq_dropna_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [],
   "source": [
    "bmg_protein_seq_df = pd.merge(protein_id_df, protein_seq_dropna_df, left_on='HGNC_Symbol', right_on='gene_name', how='left')[['BioMedGraphica_ID', 'protein_sequence']]\n",
    "bmg_protein_seq_df.to_csv('./BioMedGraphica/Entity/Protein/biomedgraphica_protein_biochem.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioMedGraphica_ID</th>\n",
       "      <th>protein_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BMG_PT000001</td>\n",
       "      <td>MGLTVSALFSRIFGKKQMRILMVGLDAAGKTTILYKLKLGEIVTTI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BMG_PT000002</td>\n",
       "      <td>MFPFYSCWRTGLLLLLLAVAVRESWQTEEKTCDLVGEKGKESEKEL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BMG_PT000003</td>\n",
       "      <td>MRTAGPTPAVSWRKRSRKRPRCPFVSYKQPAAPPSEGTRRGGAARP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BMG_PT000004</td>\n",
       "      <td>MTAEEMKATESGAQSAPLPMEGVDISPKQDEGVLKVIKREGTGTEM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BMG_PT000005</td>\n",
       "      <td>MLFEGLDLVSALATLAACLVSVTLLLAVSQQLWQLRWAATRDKSCK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205486</th>\n",
       "      <td>BMG_PT205487</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205487</th>\n",
       "      <td>BMG_PT205488</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205488</th>\n",
       "      <td>BMG_PT205489</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205489</th>\n",
       "      <td>BMG_PT205490</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205490</th>\n",
       "      <td>BMG_PT205491</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>205491 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       BioMedGraphica_ID                                   protein_sequence\n",
       "0           BMG_PT000001  MGLTVSALFSRIFGKKQMRILMVGLDAAGKTTILYKLKLGEIVTTI...\n",
       "1           BMG_PT000002  MFPFYSCWRTGLLLLLLAVAVRESWQTEEKTCDLVGEKGKESEKEL...\n",
       "2           BMG_PT000003  MRTAGPTPAVSWRKRSRKRPRCPFVSYKQPAAPPSEGTRRGGAARP...\n",
       "3           BMG_PT000004  MTAEEMKATESGAQSAPLPMEGVDISPKQDEGVLKVIKREGTGTEM...\n",
       "4           BMG_PT000005  MLFEGLDLVSALATLAACLVSVTLLLAVSQQLWQLRWAATRDKSCK...\n",
       "...                  ...                                                ...\n",
       "205486      BMG_PT205487                                                NaN\n",
       "205487      BMG_PT205488                                                NaN\n",
       "205488      BMG_PT205489                                                NaN\n",
       "205489      BMG_PT205490                                                NaN\n",
       "205490      BMG_PT205491                                                NaN\n",
       "\n",
       "[205491 rows x 2 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bmg_protein_seq_df = pd.read_csv('./BioMedGraphica/Entity/Protein/biomedgraphica_protein_biochem.csv')\n",
    "bmg_protein_seq_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_seq_df = transcript_seq_df.rename(columns={'transcript_sequence': 'sequence'})\n",
    "bmg_protein_seq_df = bmg_protein_seq_df.rename(columns={'protein_sequence': 'sequence'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioMedGraphica_ID</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BMG_TS000001</td>\n",
       "      <td>CUGCUGCUGCUGCGCCCCAUCCCCCCGCGGCCGGCCAGUUCCAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BMG_TS000002</td>\n",
       "      <td>AGAGUGGGGCACAGCGAGGCGCUAGGGGGAACGCUGGCCUCUGAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BMG_TS000003</td>\n",
       "      <td>GUCAGCUGGAGGAAGCGGAGUAGGAAGCGGCCGCGAUGUCCUUUUG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BMG_TS000004</td>\n",
       "      <td>CCUACCCCAGCUCUCGCGCCGCGUGCAGAGGUGCUCAAGCCUCCUC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BMG_TS000005</td>\n",
       "      <td>ACAGCCAAUCCCCCGAGCGGCCGCCAACAUGCUCUUUGAGGGCUUG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483812</th>\n",
       "      <td>BMG_PT205487</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483813</th>\n",
       "      <td>BMG_PT205488</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483814</th>\n",
       "      <td>BMG_PT205489</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483815</th>\n",
       "      <td>BMG_PT205490</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483816</th>\n",
       "      <td>BMG_PT205491</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>483817 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       BioMedGraphica_ID                                           sequence\n",
       "0           BMG_TS000001  CUGCUGCUGCUGCGCCCCAUCCCCCCGCGGCCGGCCAGUUCCAGCC...\n",
       "1           BMG_TS000002  AGAGUGGGGCACAGCGAGGCGCUAGGGGGAACGCUGGCCUCUGAAA...\n",
       "2           BMG_TS000003  GUCAGCUGGAGGAAGCGGAGUAGGAAGCGGCCGCGAUGUCCUUUUG...\n",
       "3           BMG_TS000004  CCUACCCCAGCUCUCGCGCCGCGUGCAGAGGUGCUCAAGCCUCCUC...\n",
       "4           BMG_TS000005  ACAGCCAAUCCCCCGAGCGGCCGCCAACAUGCUCUUUGAGGGCUUG...\n",
       "...                  ...                                                ...\n",
       "483812      BMG_PT205487                                                NaN\n",
       "483813      BMG_PT205488                                                NaN\n",
       "483814      BMG_PT205489                                                NaN\n",
       "483815      BMG_PT205490                                                NaN\n",
       "483816      BMG_PT205491                                                NaN\n",
       "\n",
       "[483817 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "s_bio_df = pd.concat([transcript_seq_df, bmg_protein_seq_df], axis=0).reset_index(drop=True)\n",
    "display(s_bio_df)\n",
    "s_bio_df.to_csv('./CellTOSG/s_bio.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Protein Emb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import List, Tuple, Dict\n",
    "from torch.utils.data import DataLoader, Dataset\n",
    "\n",
    "class SentenceDataset(Dataset):\n",
    "    def __init__(self, sentences: List[str]):\n",
    "        self.sentences = sentences\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.sentences)\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        return self.sentences[idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sequence(seq):\n",
    "    if isinstance(seq, str):\n",
    "        return seq\n",
    "    elif isinstance(seq, (list, tuple)) and len(seq) > 0:\n",
    "        return seq[0]\n",
    "    else:\n",
    "        return str(seq)  # Fallback conversion\n",
    "\n",
    "import os\n",
    "import torch\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "from typing import List\n",
    "from transformers import GPT2LMHeadModel, GPT2Tokenizer\n",
    "\n",
    "class ProtGPT_LM:\n",
    "    def __init__(self, model_path: str =\"\", model_name: str = \"prot_gpt2\", device: str = \"cpu\"):\n",
    "        \"\"\"\n",
    "        ProtGPT2-based language model.\n",
    "        Args:\n",
    "            model_name (str): Name of the ProtGPT2 model. Defaults to 'prot_gpt2'.\n",
    "            device (str): Device to run the model on ('cpu' or 'cuda'). Defaults to 'cpu'.\n",
    "        \"\"\"\n",
    "        self.model_path = model_path\n",
    "        self.model_name = model_name\n",
    "        self.device = device\n",
    "        self.model = None\n",
    "        self.tokenizer = None\n",
    "\n",
    "    def get_model(self, model_name):\n",
    "        \"\"\"\n",
    "        Initialize the ProtGPT2 model and tokenizer.\n",
    "        \"\"\"\n",
    "        model = GPT2LMHeadModel.from_pretrained(model_name).to(self.device)\n",
    "        tokenizer = GPT2Tokenizer.from_pretrained(model_name)\n",
    "        return model, tokenizer\n",
    "\n",
    "    def load_model(self):\n",
    "        \"\"\"\n",
    "        Load the ProtGPT2 model and tokenizer from the specified model name.\n",
    "        \"\"\"\n",
    "        self.model, self.tokenizer = self.get_model(self.model_name)\n",
    "        return self\n",
    "\n",
    "    def generate_embeddings(self, sequences: List[str], batch_size: int = 16, text_emb_dim: int = 64) -> np.ndarray:\n",
    "        \"\"\"\n",
    "        Generate tensor embeddings for each protein sequence with gradients enabled.\n",
    "        \n",
    "        Args:\n",
    "            sequences (List[str]): List of protein sequences.\n",
    "            batch_size (int): Batch size for processing sequences. Defaults to 16.\n",
    "            text_emb_dim (int): Dimension to which embeddings are pooled. Defaults to 64.\n",
    "        \n",
    "        Returns:\n",
    "            np.ndarray: A numpy array of shape (num_sequences, text_emb_dim) where\n",
    "                        num_sequences is the number of sequences and\n",
    "                        text_emb_dim is the dimension of the reduced embeddings.\n",
    "        \"\"\"\n",
    "        print(f\"Generating embeddings for {len(sequences)} sequences.\")\n",
    "        device = self.device\n",
    "        self.model.to(device)\n",
    "        \n",
    "        # Create a dataset and DataLoader for batching.\n",
    "        dataset = SentenceDataset(sequences)\n",
    "        # Use a custom collate_fn so the DataLoader returns the raw list of sequences.\n",
    "        dataloader = DataLoader(dataset, batch_size=batch_size, shuffle=False, collate_fn=lambda x: x)\n",
    "        \n",
    "        embeddings = []\n",
    "\n",
    "        for batch in tqdm(dataloader, desc=\"Embedding sequences\", unit=\"batch\"):\n",
    "            # Tokenize valid sequences in the batch.\n",
    "            batch_tokenized = [\n",
    "                torch.tensor(self.tokenizer.encode(get_sequence(seq)), dtype=torch.long)\n",
    "                for seq in batch if pd.notna(seq)\n",
    "            ]\n",
    "\n",
    "            if not batch_tokenized:\n",
    "                raise ValueError(\"No valid sequences found in the batch.\")\n",
    "\n",
    "            # Use the tokenizer's pad_token_id if available, otherwise default to 0.\n",
    "            pad_value = (\n",
    "                self.tokenizer.pad_token_id\n",
    "                if hasattr(self.tokenizer, \"pad_token_id\") and self.tokenizer.pad_token_id is not None\n",
    "                else 0\n",
    "            )\n",
    "\n",
    "            # Pad sequences so that they are all the same length and move the result to the target device.\n",
    "            input_ids = torch.nn.utils.rnn.pad_sequence(\n",
    "                batch_tokenized, batch_first=True, padding_value=pad_value\n",
    "            ).to(device)\n",
    "\n",
    "            # Get the model's maximum allowed sequence length.\n",
    "            max_length = self.model.config.n_positions\n",
    "\n",
    "            # Truncate padded sequences if they exceed the maximum allowed length.\n",
    "            if input_ids.size(1) > max_length:\n",
    "                input_ids = input_ids[:, :max_length]\n",
    "            \n",
    "            with torch.no_grad():\n",
    "                outputs = self.model(input_ids, labels=input_ids)\n",
    "                loss, logits = outputs[:2]\n",
    "                # Mean pooling over the token dimension => shape: [batch, hidden_dim]\n",
    "                sequence_embeddings = torch.mean(logits, dim=1)\n",
    "                # Adaptive pooling to the desired dimension (pooling over the hidden dimension)\n",
    "                projected = torch.nn.functional.adaptive_avg_pool1d(sequence_embeddings.unsqueeze(1), output_size=text_emb_dim).squeeze(1)\n",
    "                # Immediately offload to CPU and convert to numpy.\n",
    "                projected_np = projected.detach().cpu().numpy()\n",
    "                embeddings.append(projected_np)\n",
    "\n",
    "        # If embeddings were generated, concatenate them along the first dimension.\n",
    "        if embeddings:\n",
    "            embeddings_array = np.concatenate(embeddings, axis=0)\n",
    "        else:\n",
    "            embeddings_array = np.empty((0, text_emb_dim))\n",
    "            \n",
    "        print(f\"Generated embeddings shape: {embeddings_array.shape}\")\n",
    "        \n",
    "        # Convert to numpy array and return.\n",
    "        return embeddings_array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<__main__.ProtGPT_LM at 0x1a11ce78c10>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import gc\n",
    "import torch\n",
    "\n",
    "gc.collect()\n",
    "torch.cuda.empty_cache()\n",
    "device = \"cuda\"\n",
    "prot_gpt_lm = ProtGPT_LM(model_name=\"nferruz/ProtGPT2\", device=device)\n",
    "prot_gpt_lm.load_model()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "205491"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "bmg_protein_seq_df = pd.read_csv('./BioMedGraphica/Entity/Protein/biomedgraphica_protein_biochem.csv')\n",
    "protein_sequences = bmg_protein_seq_df['protein_sequence'].tolist()\n",
    "len(protein_sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "import gc\n",
    "\n",
    "torch.cuda.empty_cache()\n",
    "gc.collect()\n",
    "\n",
    "prot_gpt_lm.model.eval()  # Ensure model is in evaluation mode\n",
    "\n",
    "with torch.no_grad():\n",
    "    protein_seq_emb = prot_gpt_lm.generate_embeddings(\n",
    "        sequences=protein_sequences, \n",
    "        batch_size=4, \n",
    "        text_emb_dim=1\n",
    "    )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mkg",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
